# $$Q4$$

### Imports and File input

In [1]:
import random
import matplotlib.pyplot as plt
import numpy as np
import scipy as sp
from astropy.table import Table
from scipy.integrate import dblquad as doubleInt
from scipy.integrate import nquad
from scipy.integrate import tplquad as tripleInt

In [2]:
fileName = 'Q4.txt'
data = np.genfromtxt(fileName)

---

# $$Q4a$$ 

# These quantites apply for all the parts in the question

#### $$p(y | x, a, b) = \frac{1}{\sqrt{2\pi}\sigma_{y}} exp\left({-\frac{(y - (a+bx))^2}{2\sigma_{y}^2}}\right) $$


#### $$p(x|y, M_{1}) \propto  p(y|x)p(x) = \int_{a}\int_{b} p(y|x, a, b) p(a) p(b) \text{ db} \text{da}$$


#### $$p(a) = p(b) = \frac{1}{4}$$

In [3]:
def p_y(a, b): #this is just the function p(y|x, a, b)
    x = data[:,0]
    y = data[:,1]
    dy = data[:,2]
    p_y = (1/(np.sqrt(2*np.pi)*dy)) * np.exp(-((y - (a+(b*x)))**2)/(2 * dy**2))
    return np.prod(p_y) #the likelihood is the product of all the gaussians so I just multiplied all the elements in my list

pa = 1/4
pb = 1/4

In [4]:
a = np.linspace(-2, 2, 20)
b = np.linspace(-2, 2, 20)

In [5]:
def g(x):
    return -2
def h(x):
    return 2


ans_a = doubleInt(p_y, -2, 2, g, h)[0] * (pa * pb)

In [6]:
ans_a 

22833.18110426455

### 4.a) $p(D|M_{1}) = 22833.18110426455$

---

# $$Q4b$$ 

# These quantites apply for all the parts in the question

#### $$p(y | x, a, b, c) = \frac{1}{\sqrt{2\pi}\sigma_{y}} exp\left({-\frac{(y - (a+bx+cx^2))^2}{2\sigma_{y}^2}}\right) $$


#### $$p(x|y, M_{1}) \propto  p(y|x)p(x) = \int_{a}\int_{b} \int_{c} p(y|x, a, b, c) p(a) p(b) p(c) \text{ db} \text{da} \text{dc}$$


#### $$p(a) = p(b) = p(c) = \frac{1}{4}$$

In [7]:
def p_y(a, b, c): #this is just the function p(y|x, a, b)
    x = data[:,0]
    y = data[:,1]
    dy = data[:,2]
    p_y = (1/(np.sqrt(2*np.pi)*dy)) * np.exp(-((y - (a+(b*x)+(c*x**2)))**2)/(2 * dy**2))
    return np.prod(p_y)

pa = 1/4
pb = 1/4
pc = 1/4

In [8]:
a = np.linspace(-2, 2, 20)
b = np.linspace(-2, 2, 20)
c = np.linspace(-2, 2, 20)

In [10]:
ans = nquad(p_y, [[-2,2], [-2, 2], [-2, 2]])

In [11]:
ans_b = ans[0] * pa * pb * pc

In [12]:
ans_b

7454.563846939349

### 4.b) $p(D|M_{2}) = 7454.563846939349$

---

# $$Q4c$$ 

In [13]:
B12 = ans_a/ans_b
B12

3.0629801519024165

#### $$\text{Bayes' factor, }B_{ij} \text{ is 3.06:1 which is greater than 3:1.}\\ \text{ Therefore, we can conclude that we prefer }\\ M_{1} \text{ over } M_{2} $$

---